NullPointerException on reAuthenticate

Hi,
after session/token expiration there is a NPE on re-authenticate :

```
Caused by: ConnectionException{message=while trying to invoke the method org.openstack4j.openstack.identity.domain.v3.KeystoneAuth.unwrap() of a null object returned from org.openstack4j.openstack.identity.domain.v3.KeystoneToken.getCredentials(), status=0}
	at org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.invoke(HttpExecutorServiceImpl.java:60)
	at org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.execute(HttpExecutorServiceImpl.java:32)
	at org.openstack4j.core.transport.internal.HttpExecutor.execute(HttpExecutor.java:51)
	at org.openstack4j.openstack.internal.BaseOpenStackService$Invocation.execute(BaseOpenStackService.java:189)
	at org.openstack4j.openstack.internal.BaseOpenStackService$Invocation.execute(BaseOpenStackService.java:183)
	at org.openstack4j.openstack.compute.internal.ServerServiceImpl.get(ServerServiceImpl.java:118)
	... 56 more
Caused by: java.lang.NullPointerException: while trying to invoke the method org.openstack4j.openstack.identity.domain.v3.KeystoneAuth.unwrap() of a null object returned from org.openstack4j.openstack.identity.domain.v3.KeystoneToken.getCredentials()
	at org.openstack4j.openstack.internal.OSAuthenticator.reAuthenticate(OSAuthenticator.java:95)
	at org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.invokeRequest(HttpExecutorServiceImpl.java:71)
	at org.openstack4j.connectors.httpclient.HttpExecutorServiceImpl.invoke(HttpExecutorServiceImpl.java:56)
	... 65 more|
```
Openstack4j version is 2.0.8

Please help me solving it.
10x& best regards
Boby

--- a/core-test/src/main/java/org/openstack4j/api/AbstractTest.java
+++ b/core-test/src/main/java/org/openstack4j/api/AbstractTest.java
@@ -29,7 +29,7 @@ import com.squareup.okhttp.mockwebserver.MockWebServer;
 /**
  * Base Test class which handles Mocking a Webserver to fullfill and test against JSON response objects
  * from an OpenStack deployment
- * 
+ *
  * @author Jeremy Unruh
  */
 public abstract class AbstractTest {
@@ -55,7 +55,7 @@ public abstract class AbstractTest {
     private OSClient os;
     private String host;
     protected MockWebServer server = new MockWebServer();
-    
+
 
     /**
      * @return the service the API is using
@@ -64,10 +64,10 @@ public abstract class AbstractTest {
 
     @BeforeClass
     protected void startServer() throws UnknownHostException {
-    	
+
     	InetAddress inetAddress = InetAddress.getByName("localhost");
     	Logger.getLogger(getClass().getName()).info("localhost inet address: "+inetAddress.toString());
-    	
+
         Logger.getLogger(getClass().getName()).info("Tests using connector: " + HttpExecutor.create().getExecutorName() + " on " + getHost());
         try {
         	Logger.getLogger(getClass().getName()).info("Starting server on port "+service().port);
@@ -93,7 +93,7 @@ public abstract class AbstractTest {
         InputStream is = getClass().getResourceAsStream(resource);
         respondWith(headers, 200, new String(ByteStreams.toByteArray(is)));
     }
-    
+
     /**
      * Responds with specified status code and no body
      * @param statusCode the status code to respond with
@@ -101,7 +101,7 @@ public abstract class AbstractTest {
     protected void respondWith(int statusCode) {
         respondWith(null, statusCode, "");
     }
-    
+
     /**
      * Responds with specified status code, no body and optional headers
      * @param headers optional headers
@@ -110,7 +110,7 @@ public abstract class AbstractTest {
     protected void respondWith(Map<String,String> headers, int statusCode) {
         respondWith(headers, statusCode, "");
     }
-    
+
     /**
      * Responds with specified status code and json body
      * @param statusCode the status code to respond with
@@ -121,7 +121,7 @@ public abstract class AbstractTest {
         headers.put("Content-Type", "application/json");
         respondWith(headers, statusCode, jsonBody);
     }
-    
+
     /**
      * Responds with specified status code, body and optional headers
      * @param headers optional headers
@@ -139,7 +139,7 @@ public abstract class AbstractTest {
         r.setResponseCode(statusCode);
         server.enqueue(r);
     }
-    
+
     /**
      * Responds with given header, status code, body from json resource file
      * @param headers the specified header
@@ -149,9 +149,16 @@ public abstract class AbstractTest {
      */
     protected void respondWithHeaderAndResource(Map<String, String> headers, int statusCode, String resource) throws IOException {
         InputStream is = getClass().getResourceAsStream(resource);
-        respondWith(headers, 200, new String(ByteStreams.toByteArray(is)));
+        respondWith(headers, statusCode, new String(ByteStreams.toByteArray(is)));
     }
-    
+
+    protected void respondWithCodeAndResource(int statusCode, String resource) throws IOException {
+        InputStream is = getClass().getResourceAsStream(resource);
+        Map<String, String> headers = new HashMap<String, String>();
+        headers.put("Content-Type", "application/json");
+        respondWith(headers , statusCode, new String(ByteStreams.toByteArray(is)));
+    }
+
     protected String authURL(String path) {
         return String.format("http://%s:5000%s", getHost(), path);
     }
@@ -174,7 +181,7 @@ public abstract class AbstractTest {
 
     protected OSClient os() {
         if (os == null) {
-            ObjectMapper mapper = new ObjectMapper();	
+            ObjectMapper mapper = new ObjectMapper();
             mapper.setSerializationInclusion(Include.NON_NULL);
             mapper.enable(SerializationFeature.INDENT_OUTPUT);
             mapper.enable(SerializationFeature.WRAP_ROOT_VALUE);
@@ -193,14 +200,14 @@ public abstract class AbstractTest {
                 os = OSFactory.clientFromAccess(a);
             } catch (Exception e) {
                 e.printStackTrace();
-            } 
+            }
         }
         return os;
     }
-    
+
     protected OSClient osv3() {
         if (os == null) {
-            ObjectMapper mapper = new ObjectMapper();	
+            ObjectMapper mapper = new ObjectMapper();
             mapper.setSerializationInclusion(Include.NON_NULL);
             mapper.enable(SerializationFeature.INDENT_OUTPUT);
             mapper.enable(SerializationFeature.WRAP_ROOT_VALUE);
@@ -214,17 +221,17 @@ public abstract class AbstractTest {
                 json = json.replaceAll("127.0.0.1", getHost());
                 KeystoneToken t = mapper.readValue(json, KeystoneToken.class);
                 t.applyContext(authURL("/v3"), new Credentials("admin", "test"));
-                
+
                 //tokenwrapper
                 AccessWrapper a = AccessWrapper.wrap(t);
                 os = OSFactory.clientFromAccess(a);
             } catch (Exception e) {
                 e.printStackTrace();
-            } 
+            }
         }
         return os;
     }
-    
+
     private String getHost() {
         /*
     	try
@@ -238,7 +245,7 @@ public abstract class AbstractTest {
         */
         if (host == null)
             return "127.0.0.1";
-        
+
         return host;
     }
 }

--- a/core-test/src/main/java/org/openstack4j/api/identity/v3/KeystoneTests.java
+++ b/core-test/src/main/java/org/openstack4j/api/identity/v3/KeystoneTests.java
@@ -24,7 +24,10 @@ public class KeystoneTests extends AbstractTest {
     private static final String JSON_AUTH_DOMAINNAME_USERID = "/identity.v3/authv3_domainName_userId.json";
     private static final String JSON_AUTH_UNSCOPED = "/identity.v3/authv3_unscoped.json";
     private static final String JSON_AUTH_TOKEN = "/identity.v3/authv3_token.json";
+    private static final String JSON_AUTH_ERROR_401 = "/identity.v3/authv3_authorizationerror.json";
+    private static final String JSON_USER_LIST = "/identity.v3/usersv3_list.json";
     private static final ImmutableMap<String, String> HEADER_AUTH_PROJECT_RESPONSE = ImmutableMap.of("X-Subject-Token", "763fd7e197ab4e00b2e6e0a8d22a8e87", "Content-Type", "application/json");
+    private static final ImmutableMap<String, String> HEADER_REAUTH_PROJECT_RESPONSE = ImmutableMap.of("X-Subject-Token", "8f57cce49fd04b3cb72afdf8c0445b87", "Content-Type", "application/json");
     private static final ImmutableMap<String, String> HEADER_AUTH_TOKEN_RESPONSE = ImmutableMap.of("X-Subject-Token", "3ecb5c2063904566be4b10406c0f7568", "Content-Type", "application/json");
 
     // module test
@@ -145,11 +148,12 @@ public class KeystoneTests extends AbstractTest {
     }
 
     /**
-     * try to authenticate unscoped. should return an UnsupportedOperationException saying it's not implemented yet.
+     * try to authenticate unscoped. should return an
+     * UnsupportedOperationException saying it's not implemented yet.
      *
      * @throws Exception
      */
-    @Test(expectedExceptions={UnsupportedOperationException.class})
+    @Test(expectedExceptions = { UnsupportedOperationException.class })
     public void authenticate_unscoped_exception_Test() throws Exception {
 
         respondWith(JSON_AUTH_UNSCOPED);
@@ -193,5 +197,33 @@ public class KeystoneTests extends AbstractTest {
         assertEquals(osv3().getAccess().getUser().getId(), userId);
     }
 
+    /**
+     * reAuthenticate after authorizationError e.g. the token expired
+     *
+     * @throws Exception
+     */
+    public void reAuthentication_Test() throws Exception {
+
+        respondWithHeaderAndResource(HEADER_AUTH_PROJECT_RESPONSE, 201, JSON_AUTH_PROJECT);
+
+        associateClient(OSFactory.builderV3()
+                .endpoint(authURL("/v3"))
+                .credentials(userId, password)
+                .scopeToProject(Identifier.byId(projectId), Identifier.byId(projectDomainId))
+                .authenticate());
+
+        respondWith(JSON_USER_LIST);
+        assertEquals(os().identity().users().list().size(), 6);
+
+        // 401 Authorization error for example due to expired session..
+        respondWithCodeAndResource(401, JSON_AUTH_ERROR_401);
+
+        // .. triggers re-authentication
+        respondWithHeaderAndResource(HEADER_REAUTH_PROJECT_RESPONSE, 201, JSON_AUTH_PROJECT);
+
+        respondWith(JSON_USER_LIST);
+        assertEquals(os().identity().users().list().size(), 6);
+
+    }
 
 }

--- a/core-test/src/main/resources/identity.v3/authv3_authorizationerror.json
+++ b/core-test/src/main/resources/identity.v3/authv3_authorizationerror.json
@@ -1,0 +1,7 @@
+{
+    "error": {
+        "message": "The request you have made requires authentication. (Disable debug mode to suppress these details.)",
+        "code": 401,
+        "title": "Unauthorized"
+    }
+}

--- a/core-test/src/main/resources/identity.v3/usersv3_list.json
+++ b/core-test/src/main/resources/identity.v3/usersv3_list.json
@@ -1,0 +1,69 @@
+{
+"users": [
+    {
+        "name": "glance", 
+        "links": {
+            "self": "http://devstack.openstack.stack:35357/v3/users/07cfb841cddd46819e3dcd5df238ae04"
+            },
+        "domain_id": "default", 
+        "enabled": true, 
+        "email": null, 
+        "id": "07cfb841cddd46819e3dcd5df238ae04"
+        },
+    {
+        "name": "demo", 
+        "links": {
+            "self": "http://devstack.openstack.stack:35357/v3/users/11ba3db90590446fb6d7e0efe6c1f46a"
+            },
+        "domain_id": "default", 
+        "enabled": true, 
+        "email": "demo@example.com", 
+        "id": "11ba3db90590446fb6d7e0efe6c1f46a"
+        }, 
+    {
+        "name": "cinder", 
+        "links": {
+            "self": "http://devstack.openstack.stack:35357/v3/users/1523e8606bd1400cae36a40e0f3f817e"
+            },
+        "domain_id": "default", 
+        "enabled": true, 
+        "email": null, 
+        "id": "1523e8606bd1400cae36a40e0f3f817e"
+        },
+    {
+        "name": "nova", 
+        "links": {
+            "self": "http://devstack.openstack.stack:35357/v3/users/7afec08993c24bb09df141e513738030"
+            },
+        "domain_id": "default", 
+        "enabled": true, 
+        "email": null, 
+        "id": "7afec08993c24bb09df141e513738030"
+        },
+    {
+        "name": "u-foobar", 
+        "links": {
+            "self": "http://devstack.openstack.stack:35357/v3/users/938bd5585fd145efaadb4a7e588078c1"
+            },
+        "domain_id": "default", 
+        "enabled": true, 
+        "email": "foobar@example.com", 
+        "id": "938bd5585fd145efaadb4a7e588078c1"
+        }, 
+    {
+        "name": "admin", 
+        "links": {
+            "self": "http://devstack.openstack.stack:35357/v3/users/aa9f25defa6d4cafb48466df83106065"
+            },
+        "domain_id": "default", 
+        "enabled": true, 
+        "email": null, 
+        "id": "aa9f25defa6d4cafb48466df83106065"
+        }
+    ],
+"links": {
+    "self": "http://devstack.openstack.stack:35357/v3/users", 
+    "previous": null,
+    "next": null
+    }
+}

--- a/core/src/main/java/org/openstack4j/openstack/identity/domain/v3/KeystoneAuth.java
+++ b/core/src/main/java/org/openstack4j/openstack/identity/domain/v3/KeystoneAuth.java
@@ -18,10 +18,10 @@ import com.fasterxml.jackson.annotation.JsonRootName;
 import com.google.common.collect.Lists;
 
 /**
- * 
+ *
  * Represents an v3 Auth object.
- * 
- * 
+ *
+ *
  */
 @JsonRootName("auth")
 public class KeystoneAuth implements Authentication, AuthStore {
@@ -38,13 +38,14 @@ public class KeystoneAuth implements Authentication, AuthStore {
         this.scope = scope;
     }
 
-    public KeystoneAuth(String username, String password) {
-        this(username, password, null, null);
+    public KeystoneAuth(String userId, String password) {
+        this(userId, password, null, null);
     }
 
-    public KeystoneAuth(String username, String password, Identifier domain, AuthScope scope) {
-        this.identity = AuthIdentity.createCredentialType(username, password, domain);
+    public KeystoneAuth(String user, String password, Identifier domain, AuthScope scope) {
+        this.identity = AuthIdentity.createCredentialType(user, password, domain);
         this.scope = scope;
+        this.type = Type.CREDENTIALS;
     }
 
     protected KeystoneAuth(Type type) {
@@ -288,10 +289,12 @@ public class KeystoneAuth implements Authentication, AuthStore {
                 return domain;
             }
 
+            @Override
             public String getId() {
                 return id;
             }
 
+            @Override
             public String getName() {
                 return name;
             }
@@ -313,10 +316,12 @@ public class KeystoneAuth implements Authentication, AuthStore {
                     this.name = domain.getId();
             }
 
+            @Override
             public String getId() {
                 return id;
             }
 
+            @Override
             public String getName() {
                 return name;
             }

--- a/core/src/main/java/org/openstack4j/openstack/internal/OSAuthenticator.java
+++ b/core/src/main/java/org/openstack4j/openstack/internal/OSAuthenticator.java
@@ -168,18 +168,19 @@ public class OSAuthenticator {
         AccessWrapper accesswr = AccessWrapper.wrap(token);
         accesswr.setId(response.header(ClientConstants.HEADER_X_SUBJECT_TOKEN));
 
+        if(token.getProject() == null && token.getDomain() == null) {
+            throw new UnsupportedOperationException("Unscoped authentication is not supported in OpenStack4j 2.x . Please use OpenStack 3.y .");
+        }
+
         if (auth.getType() == Type.CREDENTIALS) {
-            token = token.applyContext(info.endpoint, new org.openstack4j.openstack.identity.domain.v3.Credentials(auth.getUsername(), auth.getPassword()));
+            token = token.applyContext(info.endpoint, auth );
         }
         else {
             if( token.getProject() != null ) {
                 token = token.applyContext(	info.endpoint, new TokenAuth(accesswr.getToken().getId(), auth.getScope().getProject().getName(), auth.getScope().getProject().getId()));
             }
-            else if( token.getDomain() != null) {
-                token = token.applyContext( info.endpoint, new TokenAuth(accesswr.getToken().getId(), auth.getScope().getDomain().getName(), auth.getScope().getDomain().getId()));
-            }
             else {
-                throw new UnsupportedOperationException("Unscoped authentication not yet supported");
+                token = token.applyContext( info.endpoint, new TokenAuth(accesswr.getToken().getId(), auth.getScope().getDomain().getName(), auth.getScope().getDomain().getId()));
             }
         }
 

